scripts/base_network.py

```diff
@@ -160,6 +157,33 @@ def _set_links_underwater_fraction(fp_offshore_shapes, n):
             )
 
 
+# TODO would be nice to join both the add_underwater* functions
+def add_underwater_lines(n, shapes):
+    if not hasattr(n.lines, "geometry"):
+        n["lines", "underwater_fraction"] = 0.0
+    else:
+        offshore_shape = gpd.read_file(shapes).unary_union
+        if offshore_shape is None or offshore_shape.is_empty:
+            n["lines", "underwater_fraction"] = 0.0
+        else:
+            lines = gpd.GeoSeries(n.lines.geometry.dropna().map(shapely.wkt.loads))
+            n.lines["underwater_fraction"] = (
+                lines.intersection(offshore_shape).length / lines.length
+            )
+
+
+def _set_links_underwater_fraction(fp_offshore_shapes, n):
+    # HVDC part always has some links as converters
+    # excluding probably purely DC networks which are currently somewhat exotic
+    if n.links.empty:
+        return
+
+    if not n.links.loc[n.links.carrier == "DC"].empty:
+        add_underwater_links(n, fp_offshore_shapes)
+    elif not n.lines.loc[n.lines.carrier == "DC"].empty:
+        add_underwater_lines(n, fp_offshore_shapes)
```

Comment:
Similarly to the case for the dc lines, do you think it may be possible to generalize the function like "add_underwater" to both lines and links?

Reply:
A very good idea! My feeling is that it can be very helpful to resolve the current troubles with functionality which are connected with `underwater_fraction` lost somewhere

Reply:
Revised. However, the current representation needs some improvements as currently it assumes that all lines are HVDC if they are located outside the onshore shape